### ISSUE TITLE
InvalidProtocolBufferException when getting bytes from Scala proto and parsing with Java proto

### ISSUE THREAD

**Issue by drexin** [scalapb/ScalaPB#391]
Because of another problem with conversion between Java and Scala protos, I tried to deserialize and serialize them from Scala to Java instead. Unfortunately that also doesn't work.

Following scenario:

I have a protobuf file with proto2 syntax. I need to convert the Scala protobuf objects to Java because of an existing API. When I get an unknown enum value from the remote service, it gets turned into an `Unrecognized` in the Scala object. Java conversions don't work, because it uses `Enum.forNumber`, which returns `null` in case it doesn't know the number, instead of returning the default value, like it would do in the regular case. So I tried converting them by serializing the Scala object and deserializing to the Java object (`Protos.Foo.parseFrom(foo.toByteArray)`. Unfortunately that also doesn't work. I get `com.google.protobuf.InvalidProtocolBufferException: Message missing required fields: foo`

When I use `syntax = proto3` it works fine. Unfortunately using proto3 syntax is not an option for us at the moment, because we depend on a protobuf definition from an external source.

**Comment by thesamet** [comment]
If I understand correctly, you are reporting two bugs:

1. Conversion from Scala to Java fails when there's an unrecognized enum.
2. ScalaPB's toByteArray gives a byte array that is not parsable by the Java protocol buffers.

Can you confirm? For (2) - can you provide a minimal protocol buffer and a Scala case class that demonstrate this problem?

**Comment by drexin** [comment]
Hey @thesamet, it's two bugs, that's correct. The Java -> Scala conversion, I already reported  while back in https://github.com/scalapb/ScalaPB/issues/357. I was able to reproduce it with the following:

**proto:**

```protobuf
syntax = "proto2";

option java_package = "scalapb.repro";
option java_outer_classname = "Protos";

enum Foo {
  BAR = 0;
  BAZ = 1;
}

message Test {
  required Foo foo = 1;
}
```

**Scala:**
```scala
package scalapb.repro

import scalapb.repro.test._

object Main {
  def main(args: Array[String]): Unit = {
    val test = Test(Foo.Unrecognized(22))

     Protos.Test.parseFrom(test.toByteArray)
  }
}
```

**Comment by thesamet** [comment]
The Java side of proto2 has some weird behavior when it comes to unrecognized enum values:

- If an enum field has a `required` label, then Java's `parseFrom` would fail for it. This is what you are reporting as (2). ScalaPB behaves differently here, it allows serializing and deserializing unrecognized enums for both proto2 and proto3. There's nothing ScalaPB can do to be able to make Java able to parse a required unrecognized enum.
- It is also impossible to instantiate a Java proto with a required enum with unrecognized value. For this reason, it is impossible to resolve (1), though there is a better error message implemented in 2825359.

For optional and repeated enum fields in proto2 it is possible to use `UnknownFieldSet` to instantiate messages with unrecognized enum values. I think that the complexity it adds to implement Java conversions for these cases is not worth it, so in 2825359 these conversions would also return an error. Even if this is implemented, it is impossible to preserve the order of a repeated list of enums when it is mixed with both recognized and unrecognized values.

Proto3 has much better support for unrecognized enums, and the commit above fixes conversions between Java and Scala.

**Comment by drexin** [comment]
@thesamet Thanks for looking into this and the explanation. I think we have an idea how to go forward now.

Also thanks a lot for your awesome work on ScalaPB!
